Recent commit about CEL causes GCC `contrib/golang` build failure

*Title*: Recent commit causes GCC build failure

*Description*:

After https://github.com/envoyproxy/envoy/commit/bd1dca3cf366bc5287295de73e4354d2f5d1f1d9 (#31456), GCC within Envoy build docker fails to build some CEL-related components and extensions. An example is given below.

```sh
git rev-parse HEAD
# bd1dca3cf366bc5287295de73e4354d2f5d1f1d9
cat user.bazelrc
# build --config=gcc
./ci/run_envoy_docker.sh 'bash'
bazel build //contrib/golang/filters/http/source:golang_filter_lib
# ERROR: /source/contrib/golang/filters/http/source/BUILD:15:17: Compiling contrib/golang/filters/http/source/golang_filter.cc failed: (Exit 1): gcc failed: error executing command (from target //contrib/golang/filters/http/source:golang_filter_lib) /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer '-std=c++0x' -MD -MF ... (remaining 221 arguments skipped)
#
# Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
# In file included from ./source/extensions/filters/common/expr/context.h:11,
#                  from ./source/extensions/filters/common/expr/evaluator.h:7,
#                  from ./contrib/golang/filters/http/source/golang_filter.h:14,
#                  from contrib/golang/filters/http/source/golang_filter.cc:1:
# external/com_google_cel_cpp/eval/public/cel_value.h:621:42: error: 'virtual absl::lts_20230802::StatusOr<const google::api::expr::runtime::CelList*> google::api::expr::runtime::CelMap::ListKeys(google::protobuf::Arena*) const' was hidden [-Werror=overloaded-virtual]
#   621 |   virtual absl::StatusOr<const CelList*> ListKeys(google::protobuf::Arena* arena) const {
#       |                                          ^~~~~~~~
# In file included from external/com_google_cel_cpp/eval/public/containers/field_backed_map_impl.h:8,
#                  from contrib/golang/filters/http/source/golang_filter.cc:26:
# external/com_google_cel_cpp/eval/public/containers/internal_field_backed_map_impl.h:46:34: note:   by 'virtual absl::lts_20230802::StatusOr<const google::api::expr::runtime::CelList*> google::api::expr::runtime::internal::FieldBackedMapImpl::ListKeys() const'
#    46 |   absl::StatusOr<const CelList*> ListKeys() const override;
#       |                                  ^~~~~~~~
# cc1plus: all warnings being treated as errors
```

The command succeeds in the previous commit.

*Relevant Links*:
https://github.com/envoyproxy/envoy/pull/31456#issuecomment-1886061775

yeah - original bug is easily reproducible

```diff
diff --git a/ci/do_ci.sh b/ci/do_ci.sh
index e500dd8792..3948a5a773 100755
--- a/ci/do_ci.sh
+++ b/ci/do_ci.sh
@@ -756,14 +756,7 @@ case $CI_TARGET in
     gcc)
         BAZEL_BUILD_OPTIONS+=("--test_env=HEAPCHECK=")
         setup_gcc_toolchain
-        echo "Testing ${TEST_TARGETS[*]}"
-        bazel_with_collection \
-            test "${BAZEL_BUILD_OPTIONS[@]}" \
-            -c fastbuild  \
-            --remote_download_minimal \
-            -- "${TEST_TARGETS[@]}"
-        echo "bazel release build with gcc..."
-        bazel_envoy_binary_build fastbuild
+        bazel build ${BAZEL_BUILD_OPTIONS[@]} //contrib/golang/filters/http/source:golang_filter_lib
         ;;
 
     info)

```

```console

# Configuration: f3b5920441742ff2e9ce1de338d1e21350b3861db729239d6673cebfb08df02f
# Execution platform: @envoy_build_tools//toolchains:rbe_linux_gcc_platform
In file included from ./source/extensions/filters/common/expr/context.h:11,
                 from ./source/extensions/filters/common/expr/evaluator.h:7,
                 from ./contrib/golang/filters/http/source/golang_filter.h:14,
                 from contrib/golang/filters/http/source/golang_filter.cc:1:
external/com_google_cel_cpp/eval/public/cel_value.h:621:42: error: ‘virtual absl::lts_20230802::StatusOr<const google::api::expr::runtime::CelList*> google::api::expr::runtime::CelMap::ListKeys(google::protobuf::Arena*) const’ was hidden [-Werror=overloaded-virtual]
  621 |   virtual absl::StatusOr<const CelList*> ListKeys(google::protobuf::Arena* arena) const {
      |                                          ^~~~~~~~
In file included from external/com_google_cel_cpp/eval/public/containers/field_backed_map_impl.h:8,
                 from contrib/golang/filters/http/source/golang_filter.cc:26:
external/com_google_cel_cpp/eval/public/containers/internal_field_backed_map_impl.h:46:34: note:   by ‘virtual absl::lts_20230802::StatusOr<const google::api::expr::runtime::CelList*> google::api::expr::runtime::internal::FieldBackedMapImpl::ListKeys() const’
   46 |   absl::StatusOr<const CelList*> ListKeys() const override;
      |                                  ^~~~~~~~
cc1plus: all warnings being treated as errors
Target //contrib/golang/filters/http/source:golang_filter_lib failed to build
INFO: Elapsed time: 79.923s, Critical Path: 77.39s
INFO: 1389 processes: 1386 remote cache hit, 3 inte
```

fwiw - trying to build a contrib bin with gcc fails with

```console
/usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -std=c++0x -fno-canonical-system-headers -Wno-builtin-macro-redefined -D__DATE__=redacted -D__TIMESTAMP__=redacted -D__TIME__=redacted -DABSL_MIN_LOG_LEVEL=4 -fdebug-types-section -fPIC -Wno-deprecated-declarations -std=c++17 -fPIC -DU_CHARSET_IS_UTF8=1 -DU_USING_ICU_NAMESPACE=0 -DUCONFIG_ONLY_HTML_CONVERSION=1 -DUCONFIG_NO_LEGACY_CONVERSION=1 -DUCONFIG_NO_BREAK_ITERATION=1 -DUCONFIG_NO_COLLATION=1 -DUCONFIG_NO_FORMATTING=1 -DUCONFIG_NO_TRANSLITERATION=1 -DUCONFIG_NO_REGULAR_EXPRESSIONS=1 -W -Wall -pedantic -Wpointer-arith -Wwrite-strings -Wno-long-long   -fuse-ld=lld -Wl,-no-as-needed -Wl,-z,relro,-z,now -B/usr/bin -pass-exit-codes -lm -fuse-ld=gold -l:libstdc++.a -Wl,--gc-sections   -o ../../bin/makeconv gencnvex.o genmbcs.o makeconv.o ucnvstat.o -L../../lib -licutu -L../../lib -licui18n -L../../lib -licuuc -L../../stubdata -licudata -lpthread -lm                                                                            makeconv.o:makeconv.cpp:DW.ref.__gxx_personality_v0: error: undefined reference to '__gxx_personality_v0'                                                                         ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function std::mutex::lock(): error: undefined reference to 'std::__throw_system_error(int)'                                            ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function std::once_flag::_Prepare_execution::~_Prepare_execution(): error: undefined reference to 'std::__once_callable'               ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function std::once_flag::_Prepare_execution::~_Prepare_execution(): error: undefined reference to 'std::__once_call'                   ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function umtx_cleanup: error: undefined reference to 'std::condition_variable::~condition_variable()'                                  ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function umtx_init::{lambda()#2}::operator()() const: error: undefined reference to 'std::condition_variable::condition_variable()'    ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function icu_72::umtx_initImplPreInit(icu_72::UInitOnce&): error: undefined reference to 'std::condition_variable::wait(std::unique_lock<std::mutex>&)'                                                                                                                                                                  ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function icu_72::umtx_initImplPostInit(icu_72::UInitOnce&): error: undefined reference to 'std::condition_variable::notify_all()'      ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function void std::call_once<void (&)()>(std::once_flag&, void (&)()): error: undefined reference to '__once_proxy'                    ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function void std::call_once<void (&)()>(std::once_flag&, void (&)()): error: undefined reference to 'std::__throw_system_error(int)'  ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function std::once_flag::_Prepare_execution::_Prepare_execution<std::call_once<void (&)()>(std::once_flag&, void (&)())::{lambda()#1}>(void (&)())::{lambda()#1}::operator()() const: error: undefined reference to 'std::__once_callable'                                                                               ../../lib/libicuuc.a(umutex.ao):umutex.cpp:function std::once_flag::_Prepare_execution::_Prepare_execution<std::call_once<void (&)()>(std::once_flag&, void (&)())::{lambda()#1}>(
void (&)()): error: undefined reference to 'std::__once_callable'                                                                                                                 
../../lib/libicuuc.a(umutex.ao):umutex.cpp:function std::once_flag::_Prepare_execution::_Prepare_execution<std::call_once<void (&)()>(std::once_flag&, void (&)())::{lambda()#1}>(
void (&)()): error: undefined reference to 'std::__once_call'                                                                                                                     
../../lib/libicuuc.a(umutex.ao):umutex.cpp:function std::unique_lock<std::mutex>::lock(): error: undefined reference to 'std::__throw_system_error(int)'                          
../../lib/libicuuc.a(umutex.ao):umutex.cpp:function std::unique_lock<std::mutex>::lock(): error: undefined reference to 'std::__throw_system_error(int)'                          
collect2: error: ld returned 1 exit status                                                                                                                                        
make[2]: *** [Makefile:81: ../../bin/makeconv] Error 1                                                                                                                            make[2]: Leaving directory '/b/f/w/bazel-out/k8-fastbuild/bin/bazel/foreign_cc/unicode_icu_build.build_tmpdir/tools/makeconv'                                                     
make[1]: *** [Makefile:47: all-recursive] Error 2                                                                                                                                 
make[1]: Leaving directory '/b/f/w/bazel-out/k8-fastbuild/bin/bazel/foreign_cc/unicode_icu_build.build_tmpdir/tools'                                                              
make: *** [Makefile:153: all-recursive] Error 2   
```

not sure if related or a new issue

--- a/bazel/cel-cpp.patch
+++ b/bazel/cel-cpp.patch
@@ -11,6 +11,54 @@ index 6aeff6d..c43864c 100644
      cel::runtime_internal::kPayloadUrlMissingAttributePath;
  
  CelValue CelValue::CreateDuration(absl::Duration value) {
+diff --git a/eval/public/containers/container_backed_map_impl.h b/eval/public/containers/container_backed_map_impl.h
+index 6092eef..7548bbf 100644
+--- a/eval/public/containers/container_backed_map_impl.h
++++ b/eval/public/containers/container_backed_map_impl.h
+@@ -30,6 +30,7 @@ class CelMapBuilder : public CelMap {
+     return values_map_.contains(cel_key);
+   }
+ 
++  using CelMap::ListKeys;
+   absl::StatusOr<const CelList*> ListKeys() const override {
+     return &key_list_;
+   }
+diff --git a/eval/public/containers/internal_field_backed_map_impl.h b/eval/public/containers/internal_field_backed_map_impl.h
+index ec773d9..caa2f3a 100644
+--- a/eval/public/containers/internal_field_backed_map_impl.h
++++ b/eval/public/containers/internal_field_backed_map_impl.h
+@@ -43,6 +43,7 @@ class FieldBackedMapImpl : public CelMap {
+   // Presence test function.
+   absl::StatusOr<bool> Has(const CelValue& key) const override;
+ 
++  using CelMap::ListKeys;
+   absl::StatusOr<const CelList*> ListKeys() const override;
+ 
+  protected:
+diff --git a/eval/public/structs/cel_proto_lite_wrap_util.cc b/eval/public/structs/cel_proto_lite_wrap_util.cc
+index 4cb21e5..35fdf7f 100644
+--- a/eval/public/structs/cel_proto_lite_wrap_util.cc
++++ b/eval/public/structs/cel_proto_lite_wrap_util.cc
+@@ -172,6 +172,7 @@ class DynamicMap : public CelMap {
+ 
+   int size() const override { return values_->fields_size(); }
+ 
++  using CelMap::ListKeys;
+   absl::StatusOr<const CelList*> ListKeys() const override {
+     return &key_list_;
+   }
+diff --git a/eval/public/structs/cel_proto_wrap_util.cc b/eval/public/structs/cel_proto_wrap_util.cc
+index fabb594..86a314c 100644
+--- a/eval/public/structs/cel_proto_wrap_util.cc
++++ b/eval/public/structs/cel_proto_wrap_util.cc
+@@ -137,6 +137,7 @@ class DynamicMap : public CelMap {
+ 
+   int size() const override { return values_->fields_size(); }
+ 
++  using CelMap::ListKeys;
+   absl::StatusOr<const CelList*> ListKeys() const override {
+     return &key_list_;
+   }
 diff --git a/internal/strings.cc b/internal/strings.cc
 index dc5a118..24457ab 100644
 --- a/internal/strings.cc

